### PR TITLE
feat: add in-transit XCM msgs in blocks endpoint

### PR DESCRIPTION
**Pull request by Imod7**
### Description
This pull request introduces the `in transit` horizontal messages within the Relay chain.

Closes issue [#732](https://github.com/paritytech/substrate-api-sidecar/issues/732) since this is the remaining [part](https://github.com/paritytech/substrate-api-sidecar/issues/732#issuecomment-1910780092). First part was completed in the [#1364](https://github.com/paritytech/substrate-api-sidecar/pull/1364) PR.

### Sample Responses

**Request**
`http://127.0.0.1:8080/blocks/17792973?decodedXcmMsgs=true`

while connected to Kusama Relay Chain

**Response**
The item `decodedXcmMsgs` now shows also the `in transit` Horizontal Messages as shown below :
```
"decodedXcmMsgs": {
    "horizontalMessages": [
      {
        "originParaId": "2084",
        "destinationParaId": "2000",
        "data": [
          {
            "v2": [
              {
                "withdrawAsset": {
          ....
           },
      {
        "originParaId": "2110",
        "destinationParaId": "2114",
        "data": [
          {
            "v2": [
              {
                "withdrawAsset": [
          ....
             ]
      }
    ],
    "downwardMessages": [],
    "upwardMessages": []
```
This block has 2 `in transit` Horizontal Messages with different origin and different destination paraId.

**Request**
`http://127.0.0.1:8080/blocks/21943991?decodedXcmMsgs=true`

while connected to Kusama Relay Chain

**Response**
The item `decodedXcmMsgs` shows 2 `in transit` Horizontal Messages with same origin and same destination paraId: 
```
"decodedXcmMsgs": {
    "horizontalMessages": [
      {
        "originParaId": "2001",
        "destinationParaId": "2023",
        "data": [
          {
            "v3": [
              {
                "withdrawAsset": [
                  {
                   ....
          },
          {
            "v3": [
              {
                "withdrawAsset": [
                  {
                    ....
      }
    ],
    "downwardMessages": [],
    "upwardMessages": []
  }
```

*** Same response structure when connected to a Parachain, e.g. block 5646401 from Polkadot Asset Hub

**Request**
`http://127.0.0.1:8080/blocks/22238422?decodedXcmMsgs=true`

while connected to Kusama Relay Chain

**Response**
The item `decodedXcmMsgs` shows 3 Upward Messages from same origin paraId :
```
 "decodedXcmMsgs": {
    "horizontalMessages": [],
    "downwardMessages": [],
    "upwardMessages": [
      {
        "originParaId": "2001",
        "data": [
          {
            "v3": [
              {
                "withdrawAsset": [
                 ....
          {
            "v3": [
              {
                "withdrawAsset": [
                  {
                   ....
          },
          {
            "v3": [
              {
                "withdrawAsset": [
                  {
                  ...
    ]
  }
```

### Tested Cases with Example Blocks
Cases that were tested (while connected to a Kusama/Polkadot Relay chain): 
1. Blocks with 1 Horizontal Message.
1. Blocks with 1 Horizontal and 1 Upward Message.
1. Blocks with 2 Horizontal Messages with same origin paraId and different destination paraId.
1. Blocks with 2 Horizontal Messages with different origin and different destination paraId.
2. Blocks with multiple Upward Messages with same origin paraId .

*** Above cases were also tested with the query param `paraId`

All examples tested can be found [here](https://github.com/Imod7/testing-sidecar/blob/main/decoded-xcm-msg/blocks-with-xcm-msgs/README.md)

## Changelog (with Fixes - Breaking Changes)
- ❗In Upward Messages of Relay, `data` is now an `array` since multiple XMC messages from same paraId can be retrieved.
- ❗The structure/shape (and consequently the associated `interface`) of Horizontal Messages in the **Relay** Chain differs from that of Horizontal Messages in **Parachain**. As a result:
	- `IHorizontalMessageInRelayChain` interface was introduced.
	- `IHorizontalMessage` was renamed to `IHorizontalMessageInParachain` to provide clearer distinction.
- ❗The field `paraId` was renamed to `originParaId` in the following interfaces / responses:
	- `IHorizontalMessageInParachain`
	- `IUpwardMessage`
	to explicitly indicate that it refers to the **origin** paraId rather than the **destination**. It also ensures consistency with the `IHorizontalMessageInRelayChain` interface.
- The code has been refactored:
	- when checking for Upward and Horizontal Messages in Relay and
	- when checking for Horizontal Msgs in Parachain
	to avoid repetitive blocks of code
- Corrected `ISanitizedParachainInherentData` so that it reflects how the information is retrieved.

### Todos
- [x] Update Documentation

### CODE REVIEW

**Review by marshacb — APPROVED** [review]
Solid PR. LGTM!

**Review by IkerAlus — CHANGES_REQUESTED** [review]
As discussed in the call, the output should something like:

When connected in Relay and I have 5 Horizontal Msgs like
- 4 from para `2010`, 2 to para `2034`, 1 to para `2031`, 1 to para `2032`.
- 1 from para `2020` to para `2032`,

 they should be shown as:

```
"decodedXcmMsgs": {
    "horizontalMessages": [
      {
        "sentAt": "blockNumber",
        "originId":"2010",
        "destinationId": "2034",
        "data":           {
            "v3": [
              {
                "withdrawAsset": [
                  {...
            ]
          },
          {
            "v3": [
              {
                "withdrawAsset": [
                  {
                 ...
           ]
        },
        {
        "sentAt": "blockNumber",
        "originId":"2010",
        "destinationId": "2031",
        "data": [
          {
            "v3": [
              {
                "withdrawAsset": [
                  {
                ...
        },
        {
        "sentAt": "blockNumber",
        "originId":"2010",
        "destinationId": "2032",
        "data": [
          {
            "v3": [
              {
                "withdrawAsset": [
                  {
                ...
        },
        {
        "sentAt": "blockNumber",
        "originId":"2020",
        "destinationId": "2032",
        "data": [
          {
            "v3": [
              {
                "withdrawAsset": [
                  {
                ...
          }
        ]
      }
```

Perhaps there are better choices for the naming of `originId` and `destinationId` but the structure should be like the one above.

File: src/services/blocks/XCMDecoder.ts [hunk 1]
```diff
@@ -100,44 +107,67 @@ export class XcmDecoder {
...
-							if (paraId !== undefined && index.toString() === paraId.toString()) {
-								horizontalMessage = {
-									sentAt: msg.sentAt,
-									paraId: index,
-									data: xcmMessageDecoded,
-								};
-								xcmMessages.horizontalMessages?.push(horizontalMessage);
-							} else if (paraId === undefined) {
-								horizontalMessage = {
+						if (paraId === undefined || index.toString() === paraId.toString()) {
```
Comment (TarikGul) [inline src/services/blocks/XCMDecoder.ts:110]:
```suggestion
						if (!paraId || index.toString() === paraId.toString()) {
```

File: src/services/blocks/XCMDecoder.ts [hunk 2]
```diff
@@ -100,44 +107,67 @@ export class XcmDecoder {
...
+			const paraIdCandidate: string = candidate.candidate.descriptor.paraId.toString();
+
+			xcmMessages.forEach((msg: string | ISanitizedBackedCandidateHorizontalMessage) => {
+				const msgData: string =
+					messageType === 'upward'
+						? (msg as string)
+						: (msg as ISanitizedBackedCandidateHorizontalMessage).data.slice(1);
+				const xcmMessageDecoded: string = this.decodeMsg(api, msgData);
+
+				if (paraId === undefined || paraIdCandidate === paraId.toString()) {
```
Comment (TarikGul) [inline src/services/blocks/XCMDecoder.ts:150]:
```suggestion
				if (!paraId || paraIdCandidate === paraId.toString()) {
```

**Review by TarikGul — APPROVED** [review]
LGTM small nits

**Review by IkerAlus — APPROVED** [review]
LGTM, good job!

I see there is no "sentAt" for horizontal messages in the Relay chain but I think that in this context it is not so relevant, I am ok with leaving as it is now.

Also, consider the doc update so it is clear to the user what we are showing in each of the cases.

**Comment by Imod7** [comment]
> LGTM, good job!
> 
> I see there is no "sentAt" for horizontal messages in the Relay chain but I think that in this context it is not so relevant, I am ok with leaving as it is now.
> 
> Also, consider the doc update so it is clear to the user what we are showing in each of the cases.

Yes, there is no "sentAt" field available so that is why I didn't put it. Yes, definitely I will update the docs to make everything very clear and also the description of the PR with the latest changes. Thank you!

File: docs/src/openapi-v1.yaml [hunk 3]
```diff
@@ -2997,21 +2994,63 @@ components:
...
+        type: object
+        properties:
+          originParaId:
+            type: string
+            format: unsignedInteger
+            description: The Parachain id that the specific XCM message was sent from.
+          destinationParaId:
+            type: string
+            format: unsignedInteger
+            description: The Parachain id that the specific XCM message was sent at.
```
Comment (IkerAlus) [inline docs/src/openapi-v1.yaml:3028]: I think the better form is "sent to".

Comment (IkerAlus) [inline docs/src/openapi-v1.yaml:3028]:
```suggestion
            description: The Parachain id that the specific XCM message was sent to.
```

**Pull request merged**